src/arithmetic/path_funcs/path_funcs.c

```diff
@@ -0,0 +1,179 @@
+	// If there are edges in the path, verify the node is referenced in the last edge.
+#ifndef NDEBUG
+	if(edge_count > 0) {
+		Edge *e = &path->edges[edge_count - 1];
+		EntityID nodeId = ENTITY_GET_ID(node);
+		assert(nodeId == e->srcNodeID || nodeId == e->destNodeID);
+	}
+#endif
+}
+
+/**
+ * @brief  Reverse the direction of an edge.
+ * @note   A new copy of the edge is created, in order not to invalidate any data in the original edge
+ *         in case it required elsewhere.
+ * @param  *e: Pointer to the original edge.
+ * @retval New edge with inverse direction.
+ */
+static Edge *_Edge_ReverseDirection(Edge *e) {
+	Edge *clone = Edge_Clone(e);
+	Edge_SetSrcNode(clone, e->dest);
+	Edge_SetDestNode(clone, e->src);
+	return clone;
+}
+
+/**
+ * @brief  Appends an edge to path object during path build.
+ * @note   The edge should be added after its source node has been inserted to its right position in the path.
+ *         Edges insertion is done by interliving nodes and edges.
+ * @param  path: Path.
+ * @param  e: Edge.
+ * @param  RTLEdge: Indicates the if the edge is incoming (RTL in query) edge or not.
+ * @retval None
+ */
+static void _PathBuilder_AppendEdge(Path *path, Edge *edge, bool RTLEdge) {
+	assert(Path_NodeCount(*path) > 0);
+	// The edge should connect nodes[edge_count] to nodes[edge_count+1]
+	uint edge_count = Path_EdgeCount(*path);
+	Node n = (*path).nodes[edge_count];
+	EntityID nId = ENTITY_GET_ID(&n);
+	// Validate source node is in the right place.
+	assert(nId == edge->srcNodeID || nId == edge->destNodeID);
+	/* Reverse direction if needed. A direction change is needed if the last node in the path, reading
+	 * RTL is the source node in the edge, and the edge direction in the query is LTR.
+	 * path =[(a)]
+	 * e = (a)->(b)
+	 * Query: MATCH p=(a)<-[]-(b)
+	 * e direction needs to be change. */
+
+	if(RTLEdge && nId == edge->srcNodeID) edge = _Edge_ReverseDirection(edge);
+	Path_AppendEdge(path, *edge);
+}
+
+static void _PathBuilder_AppendPath(Path *path, Path *new_path, bool RTLEdge) {
+	uint path_node_count = Path_NodeCount(*path);
+	assert(path_node_count > 0);
+	// No need to append empty paths.
+	uint new_path_node_count = Path_NodeCount(*new_path);
+	if(new_path_node_count == 0) return;
+
+	Node *last_LTR_node = &path->nodes[path_node_count - 1];
+	EntityID last_LTR_node_id = ENTITY_GET_ID(last_LTR_node);
+	Node *new_path_node_0 = &new_path->nodes[0];
+	EntityID new_path_node_0_id = ENTITY_GET_ID(new_path_node_0);
+	Node *new_path_last_node = &new_path->nodes[new_path_node_count - 1];
+	EntityID new_path_last_node_id = ENTITY_GET_ID(new_path_last_node);
+	// Validate current last LTR node is in either edges of the path.
+	assert(last_LTR_node_id == new_path_node_0_id || last_LTR_node_id == new_path_last_node_id);
+	int new_path_edge_count = Path_EdgeCount(*new_path);
+	// Check if path needs to be rverse inserated or not.
+	if(last_LTR_node_id == new_path_last_node_id) {
```

Comment:
In this case wouldn't you prefer to simply reverse the entire new_path
then the concatenation logic is the same for both reversed and not reversed paths?

Reply:
done